File: Advent_of_code_2025_Python/day7/day7.py
"""
Author: Matthieu Borelle
Description: Solution to Day 7 of the Advent of Code 2025 challenge.
"""

import numpy as np
import math
import re
from pathlib import Path


SCRIPT_DIR = Path(__file__).parent

def load_data(file_name):
    """Load data from a text file into arrays."""
    with open(SCRIPT_DIR / file_name) as f:

        lines = [line.strip('\n') for line in f]
            # # Split each line into a list of integers   

    return lines



## Part 1
def part1_solution(data_file_name):
    """
    Returns the solution of the part1 of the day7 challenge as an integer

    The number of tachion beam splitted
    """
    lines = load_data(data_file_name)

    #print(lines)   

    n_c = len(lines[0])
    n_l = len(lines)

    sum_beam_splitted = 0

    # # detect tachyon beam enters (S)
    
    list_beam = [[] for i in range(n_l)] #[]

    for j in range(0,n_c-1):

        if lines[0][j] == "S":
             list_beam[0].append(j)

    #print(list_beam)


    for i in range(1,n_l):

        for elt in list_beam[i-1]:
            
            if lines[i][elt] == "^": # beam splitting
                
                sum_beam_splitted += 1

                if elt>= 1:
                    list_beam[i].append(elt-1)
                if elt<n_c:
                    list_beam[i].append(elt+1)


            else:   # vertical beam propagation

                list_beam[i].append(elt)


            # remove plural number redundancies
            list_beam[i] = list(dict.fromkeys(list_beam[i]))

  
    #print("Dim l*c: ",n_l,n_c)
    #print(list_beam)

    return sum_beam_splitted


# def display the beam propagation(lines,list_beam): 

## Part 2


def part2_solution(data_file_name):
    """
    Returns the solution of the part2 of the day7 challenge as an integer

    The number of tachion beam timelines
    """
    lines = load_data(data_file_name)

    #print(lines)   

    n_c = len(lines[0])
    n_l = len(lines)


    # # detect tachyon beam enters (S)
    
    list_number_beams = [[0]*n_c for i in range(n_l)] #[]

    for j in range(0,n_c-1):

        if lines[0][j] == "S":
             list_number_beams[0][j] += 1

    #print(list_number_beams)


    for i in range(1,n_l):

        #print(i,"/",n_l)
        for elt in range(n_c):
            
            if list_number_beams[i-1][elt] > 0:
            
                if lines[i][elt] == "^": # beam splitting
                    

                    if elt>=1:
                        list_number_beams[i][elt-1] += list_number_beams[i-1][elt] 
                    if elt<n_c:
                        list_number_beams[i][elt+1] += list_number_beams[i-1][elt]


                else:   # vertical beam propagation

                    list_number_beams[i][elt] += list_number_beams[i-1][elt]


            # remove plural number redundancies
            #list_beam[i] = list(dict.fromkeys(list_beam[i]))

    #print("Dim l*c: ",n_l,n_c)
    #print(list_number_beams[n_l-1])

    number_timelines = sum(list_number_beams[n_l-1])

    return number_timelines
    


if __name__ == "__main__": 

    print("Part 1 test:", part1_solution("day7_input_test.txt"))   # 21
    # #print("Part 1 test:", part1_solution("day7_input_test2.txt")) # 
    print("Part 1 solution:", part1_solution("day7_input.txt")) # 1594
    print("---")
    print("Part 2 test:", part2_solution("day7_input_test.txt"))  # 40
    #print("Part 2 test:", part2_solution("day7_input_test2.txt")) 
    print("Part 2 solution:", part2_solution("day7_input.txt")) # 15650261281478
    

File: Advent_of_code_2025_C++/day7/day7.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <unordered_set>

using namespace std;

// Function to load data from a text file
vector<string> load_data(const string& file_name) {
    ifstream file(file_name);
    vector<string> lines;
    string line;
    while (getline(file, line)) {
        lines.push_back(line);
    }
    return lines;
}

// Part 1 solution
int part1_solution(const string& data_file_name) {
    vector<string> lines = load_data(data_file_name);

    int n_c = lines[0].size();
    int n_l = lines.size();

    int sum_beam_splitted = 0;

    // Initialize list_beam
    vector<vector<int>> list_beam(n_l);

    // Detect tachyon beam enters (S)
    for (int j = 0; j < n_c; ++j) {
        if (lines[0][j] == 'S') {
            list_beam[0].push_back(j);
        }
    }

    for (int i = 1; i < n_l; ++i) {
        unordered_set<int> unique_beams;
        for (int elt : list_beam[i-1]) {
            if (lines[i][elt] == '^') { // beam splitting
                sum_beam_splitted += 1;
                if (elt >= 1) {
                    unique_beams.insert(elt - 1);
                }
                if (elt < n_c - 1) {
                    unique_beams.insert(elt + 1);
                }
            } else { // vertical beam propagation
                unique_beams.insert(elt);
            }
        }
        // Convert unordered_set to vector
        list_beam[i].assign(unique_beams.begin(), unique_beams.end());
    }

    return sum_beam_splitted;
}

// Part 2 solution
long long part2_solution(const string& data_file_name) {
    vector<string> lines = load_data(data_file_name);

    int n_c = lines[0].size();
    int n_l = lines.size();

    // Initialize list_number_beams
    vector<vector<long long>> list_number_beams(n_l, vector<long long>(n_c, 0));

    // Detect tachyon beam enters (S)
    for (int j = 0; j < n_c; ++j) {
        if (lines[0][j] == 'S') {
            list_number_beams[0][j] += 1;
        }
    }

    for (int i = 1; i < n_l; ++i) {
        for (int elt = 0; elt < n_c; ++elt) {
            if (list_number_beams[i-1][elt] > 0) {
                if (lines[i][elt] == '^') { // beam splitting
                    if (elt >= 1) {
                        list_number_beams[i][elt-1] += list_number_beams[i-1][elt];
                    }
                    if (elt < n_c - 1) {
                        list_number_beams[i][elt+1] += list_number_beams[i-1][elt];
                    }
                } else { // vertical beam propagation
                    list_number_beams[i][elt] += list_number_beams[i-1][elt];
                }
            }
        }
    }

    long long number_timelines = 0;
    for (long long num : list_number_beams[n_l-1]) {
        number_timelines += num;
    }

    return number_timelines;
}

int main() {
    cout << "Part 1 test: " << part1_solution("day7_input_test.txt") << endl;   // 21
    cout << "Part 1 solution: " << part1_solution("day7_input.txt") << endl;   // 1594
    cout << "---" << endl;
    cout << "Part 2 test: " << part2_solution("day7_input_test.txt") << endl;  // 40
    cout << "Part 2 solution: " << part2_solution("day7_input.txt") << endl; // 15650261281478
    return 0;
}

